### DISCUSSION TITLE
Shifted 24 hour interval for hot water boiler

### DISCUSSION THREAD

**Discussion by aherbjornsen** [erlendsellie/priceanalyzer#24] [Ideas]
I have noticed that the current implementation for hot water boiler is not  optimal.

In general I want to heat water during the night, both because it is cheaper and because we often shower in the morning.

But the current calculations for cheapest hours have a cutoff point at midnight, and certain days (when the price is dropping) some of the 5 cheapest hours are late evening before midnight. A couple of days ago I even had "cheapest hour" between 23:00 and 24:00.

But I do not want to  heat my hot water to 80 degrees that early, because the price is (most likely) even lower 3-4 hours later.

I did fix this sort of by adding a template sensor to avoid excessive heating late in the day.

Would it be possible to do the calculations for cheapest hours for the time span from mid-day to mid-day? I.e. shifting them 12 hours.

**Comment by erlendsellie** [comment]
Hi!
Sorry, i did not see any notification about this discussion, and just stumbled upon it now. 
I can see the issue. I've noticed the same issue when the price is getting lower the next day, and PriceAnalyzer will heat up the heater around 22-24, just before the price will go further down. I've got a fix for this in my next release, but it has not solved your issue.

Shifting the day 12 hours would do the trick, and is a great suggestion, but implementing it would possibly require a lot of rewrite to the integration, as both the integration and Nordpool is working with 'days' as set constant.

My latest code is analyzing if an hour is 'Cheap compared to the future' or 'Is low compared to tomorrow', and will boost the heater extra, if it is. I may be able to flip this logic as well, and probably solve the issue you are referring to this way.  Does that sound like a fix?

Here's a visulized exampled of logic i've already made:
![image](https://user-images.githubusercontent.com/18458417/220151896-454cc45b-c136-41ba-a729-e362cd5ce086.png)


Thanks for the suggestion!

**Reply by aherbjornsen** [reply]
Sorry for the late reply, I've been on vacation...
Inverting the the "future" logic would go a long way, though I've lately seen some cases lately with the cheapest hours being aroung mid-day. So in order to make sure I get hot water in the morning I guess I will have to do some templating, finding the cheapest hours at night.

**Reply by erlendsellie** [reply]
> Sorry for the late reply, I've been on vacation... Inverting the the "future" logic would go a long way, though I've lately seen some cases lately with the cheapest hours being aroung mid-day. So in order to make sure I get hot water in the morning I guess I will have to do some templating, finding the cheapest hours at night.

I'll see what i can do about this. With my config for the hot water heater (with temperature probe on the heater) i always have enough hot water for a shower.
Do you use it as a binary sensor so it just cut's power for X amount of hours?

This template may be of some use, maybe you can modify it some to achive what you are looking for.
It picks out the 5 cheapest hours between now, and before 0700 tomorrow.
https://github.com/erlendsellie/priceanalyzer/wiki/Additional-templating#find-the-five-cheapest-hours-from-now-and-0700-tomorrow

**Reply by aherbjornsen** [reply]
Thanks, I'll look at the template and will probably be able to adjust it to my needs.

I use the sensor as input to a thermostat, but with the current price development, it does not work as I want.

I  need a full hot water tank in the morning, so I do want it to heat to 70 or 80 degrees some time during the night. 

Currently the price fluctuations are very small, and the cheapest hours seem to be around mid-day. That does not work very well. For today I see the values:

12:00 - 70
13:00 - 70
14:00 - 80
etc.

And then it stays at 40 and 50 degrees during the night.

So what I need is probably to pick the 2-3 cheapest hours between 22:00 and 05:00 and heat to 70 or 80 degress in that time period.
